static/hub.ts

```diff
@@ -0,0 +1,381 @@
+            elem = elem.parent;
+        }
+        return elem;
+    }
+
+    public findEditorInChildren(elem: GoldenLayout.ContentItem): GoldenLayout.ContentItem | boolean {
+        const count = elem.contentItems.length;
+        let index = 0;
+        while (index < count) {
+            const child = elem.contentItems[index];
+            // @ts-expect-error -- GoldenLayout's types are messed up here?, this is
+            // valid, and componentName will be present on all registered Components
+            if (child.componentName === 'codeEditor') {
+                return this.findParentRowOrColumnOrStack(child);
+            } else {
+                if (child.isRow || child.isColumn || child.isStack) {
+                    const editor = this.findEditorInChildren(child);
+                    if (editor) return editor;
+                }
+            }
+            index++;
+        }
+        return false;
+    }
+
+    public findEditorParentRowOrColumn(): GoldenLayout.ContentItem {
+        return this.findParentRowOrColumn(this.layout.root);
+    }
+
+    public addInEditorStackIfPossible(elem: GoldenLayout.ContentItem): void {
+        const insertionPoint = this.findEditorParentRowOrColumn();
+        if (insertionPoint) {
+            insertionPoint.addChild(elem);
+        } else {
+            this.addAtRoot(elem);
+        }
+    }
+
+    public addAtRoot(elem: GoldenLayout.ContentItem): void {
+        const rootFirstItem = this.layout.root.contentItems[0];
+        if (rootFirstItem) {
+            if (rootFirstItem.isRoot || rootFirstItem.isColumn) {
+                rootFirstItem.addChild(elem);
+            } else {
+                // @ts-expect-error -- GoldenLayout's types are messed up here?
+                const newRow: ContentItem = this.layout.createContentItem({
+                    type: 'row',
+                }, this.layout.root);
+                this.layout.root.replaceChild(rootFirstItem, newRow);
+                newRow.addChild(rootFirstItem);
+                newRow.addChild(elem);
+            }
+        } else {
+            this.layout.root.addChild({
+                type: 'row',
+                content: [elem],
+            });
+        }
+    }
+
+    public activateTabForContainer(container?: GoldenLayout.Container) {
+        if (container && container.tab) {
+            container.tab.header.parent.setActiveContentItem(container.tab.contentItem);
+        }
+    }
+
+    // Component Factories
+
+    private codeEditorFactory(container: GoldenLayout.Container, state: any): void {
+        // Ensure editors are closable: some older versions had 'isClosable' false.
+        // NB there doesn't seem to be a better way to do this than reach into the config and rely on the fact nothing
+        // has used it yet.
+        container.parent.config.isClosable = true;
+        const editor = new Editor(this, state, container);
+        this.editors.push(editor);
+    }
```

Comment:
This returns void, but in the layout.registerComponent, we are attempting to return it, possible bug?